### ISSUE TITLE
Observed Behavior: pyodbc.drivers() not thread safe

### ISSUE THREAD

**Issue by ourcookiemonster** [mkleehammer/pyodbc#715]
pyodb.drivers() method can return an incomplete list of drivers if multiple threads call this method at the same time.

Based on the ODBC SQLDrivers method API, the c++ code must iterate through a list of drivers, specifying the start or the next available driver.  If two threads have entered the for loop of mod_drivers() in pyodbcmodule.cpp one or both of the threads may miss a driver that the other thread has incremented through.

**Comment by ourcookiemonster** [comment]
It is likely mod_datasources() suffers the same behavior.

**Comment by ourcookiemonster** [comment]
import threading
import pyodbc

true_driver_count = len(pyodbc.drivers())
disparity_found = False

def test_drivers():
    global disparity_found
    while not disparity_found:
        thread_driver_count = len(pyodbc.drivers())
        if(thread_driver_count < true_driver_count):
            print('Driver Count Difference Detected')            
            disparity_found = True;

if __name__ == "__main__":
    thread_one = threading.Thread(target=test_drivers)
    thread_two = threading.Thread(target=test_drivers)

    thread_one.start()
    thread_two.start()

    thread_one.join()
    thread_two.join()

**Comment by ourcookiemonster** [comment]
`import threading
import pyodbc

true_driver_count = len(pyodbc.drivers())
disparity_found = False

def test_drivers():
    global disparity_found
    while not disparity_found:
        thread_driver_count = len(pyodbc.drivers())
        if(thread_driver_count < true_driver_count):
            print('Driver Count Difference Detected')            
            disparity_found = True;

if __name__ == "__main__":
    thread_one = threading.Thread(target=test_drivers)
    thread_two = threading.Thread(target=test_drivers)

    thread_one.start()
    thread_two.start()

    thread_one.join()
    thread_two.join()`

**Comment by ourcookiemonster** [comment]
```
import threading
import pyodbc

true_driver_count = len(pyodbc.drivers())
disparity_found = False

def test_drivers():
    global disparity_found
    while not disparity_found:
        thread_driver_count = len(pyodbc.drivers())
        if(thread_driver_count < true_driver_count):
            print('Driver Count Difference Detected')            
            disparity_found = True;

if __name__ == "__main__":
    thread_one = threading.Thread(target=test_drivers)
    thread_two = threading.Thread(target=test_drivers)

    thread_one.start()
    thread_two.start()

    thread_one.join()
    thread_two.join()
```

**Comment by v-makouz** [comment]
I can reproduce and will look into it

**Comment by v-makouz** [comment]
I made a PR that should fix it:
https://github.com/mkleehammer/pyodbc/pull/722

I ran your repro and after several million iterations there were no mismatches

**Comment by keitherskine** [comment]
Fixed by #722.
